### DISCUSSION TITLE
The DS1302 library examples didn't work in Arduino IDE.

### DISCUSSION THREAD

**Discussion by mianqi2016** [Makuna/Rtc#135] [Q&A] [answered]
I tested your DS1302 in Arduino IDE, the "ds1302" and "datecalc" were all not work,

the former prompted:

'DS1302' does not name a type,

the latter:
'const class DateTime' has no member named 'tostr'

Sorry,  I confused this library with another "RTCLib by NeiroN", the prompt was:RTC lost confidence in the DateTime!

**Comment by Makuna** [comment, marked as answer]
What Arduino Board are you using?

What sketch are you using?  
Include at least several lines of the serial output.

The most common cases of this message are the pins are not connected correctly and/or the onboard battery is dead.   
https://github.com/Makuna/Rtc/wiki/FAQ-%235

**Comment by mianqi2016** [comment]
I found the reason:

`
//ThreeWire myWire(4,5,2); // IO, SCLK, CE

ThreeWire myWire(17,16,26); // IO, SCLK, CE

// Since the board in experiment was Espduino

// (Arduino UNO based on ESP32)

// so pin-numbers according to Arduino UNO(4,5,2)

// are IO(17,16,26)
`

**Reply by mianqi2016** [reply]
Now the problem was:
07:23:25.748 -> 01/09/2000 83:83:00
07:23:25.748 -> RTC lost confidence in the DateTime!

and, after I click "RESET":

> 07:24:25.775 -> 01/09/2000 83:00:00
07:24:25.775 -> RTC lost confidence in the DateTime!
07:24:28.931 -> ⸮⸮
HHZJJJ⸮8⸮!⸮⸮⸮⸮compiled: Jan 19 202207:19:34
07:24:29.031 -> 01/19/2022 07:19:34
07:24:29.084 -> RTC lost confidence in the DateTime!
07:24:29.084 -> RTC is older than compile time!  (Updating DateTime)
07:24:29.084 -> 01/09/2000 07:09:08
07:24:39.057 -> 01/09/2000 83:09:00
07:24:39.057 -> RTC lost confidence in the DateTime!
07:24:49.037 -> 01/09/2000 83:09:00

**Reply by mianqi2016** [reply]
And I tried "DS1307_Simple", it was OK.

> 07:31:58.448 -> 01/19/2022 07:31:45
07:32:08.428 -> 01/19/2022 07:31:55
07:32:18.455 -> 01/19/2022 07:32:05
07:32:28.448 -> 01/19/2022 07:32:15
07:32:38.456 -> 01/19/2022 07:32:25
07:32:48.419 -> 01/19/2022 07:32:35
07:32:58.450 -> 01/19/2022 07:32:45
07:33:08.424 -> 01/19/2022 07:32:55

**Reply by Makuna** [reply]
Are you sure you have a DS1302?  The two are not the same and you state that SD1307 works for you.

The DS1302 has three connections, Clock (CLK/SCLK), Data (DAT/IO), and Chip Enable (RST/CE).  This is why it uses the ThreeWire object (from my library).  The ThreeWire object is bitbang, so it can use any available IO pins.

The DS1307 has only two connections, Clock (SCL) and Data (SDA).  This is why it uses either TwoWire or SoftwareWire (both Arduino standard objects).

**Reply by mianqi2016** [reply]
I'm sure I used DS1302, since I inserted one DS1307 and one DS1302 side by side on breadboard: https://ibb.co/XYmcyVf

**Reply by Makuna** [reply]
Please be more specific and state details.  
I confirmed it worked the RTC module I have that has a DS1302 on it with a non-ESP32 board I had handy.
Since you are not using an already wired module, I suspect your wiring might be causing issues.
Make sure the GNDs are all connected together.
Make sure the VCCs match (3.3v for ESP32).
Confirm the pins you are using are not conflicting with anything.  The ESP32 exposes pins that shouldn't be used and even some pins that are limited in functionality; check with the ESP32 specs.

**Reply by mianqi2016** [reply]
I saw the example used pin 2, is it necessary for UNO interrupt, or I can alter it to other pins?

**Reply by Makuna** [reply]
The ThreeWire is bitbang, no hardware used, no interrupts used.

**Reply by mianqi2016** [reply]
I tried:

> ThreeWire myWire(6,7,5); // IO, SCLK, CE
ThreeWire myWire(5,6,4); // IO, SCLK, CE
ThreeWire myWire(4,5,2); // IO, SCLK, CE

on Arduino UNO, both of them shown: https://ibb.co/LgDcg2y

**Reply by mianqi2016** [reply]
show me your module, I want to see its schematic.

**Reply by Makuna** [reply]
![image](https://user-images.githubusercontent.com/7972357/150592659-0d9c602c-5b45-442c-81df-6324e20a14c3.png)
I will be connecting it to a ESP32 in next few days to see if it something about an ESP32.

**Reply by mianqi2016** [reply]
It looks totally the same with my wiring on breadboard - no any resistors and capacitors, just wire directly connection.

**Reply by mianqi2016** [reply]
I used a new components of breadborad+DS1302+Crystal and then it was OK, thanks for your answer to my issue. I think you were right - " the pins are not connected correctly " although I haven't start to find why the former group didn't work. Your code is clear and would be more useful when added the weekday info. Thanks.

**Reply by Makuna** [reply]
Just confirmed it works with ESP32.
```
ThreeWire myWire(17,16,5); // IO, SCLK, CE
```
This equates to in order, as labeled on my ESP32 Dev board; TX2, RX2, and D5.
Note my RTC module labels CE as RST.
ESP boards sometimes label D pins that don't match their GPIO number.

Also note, the spec sheet for the chip states to make sure the crystal is backed by GND plane and the wires leading to it as also well shielded.

**Reply by mianqi2016** [reply]
I tried it just now, it works:
ThreeWire myWire(27,16,17); // IO, SCLK, CE

**Reply by mianqi2016** [reply]
![image](https://user-images.githubusercontent.com/22947586/150898230-8c8974e2-8b83-46e4-9948-5ea78f02d36b.png)

**Reply by mianqi2016** [reply]
![image](https://user-images.githubusercontent.com/22947586/150898812-5c6ccf35-b193-4ee9-a750-4b7e21108487.png)
